### ISSUE TITLE
Heap types implemented in C must have Py_TPFLAGS_HAVE_GC and implement tp_traverse

### ISSUE THREAD

**Issue by vstinner** [python/cpython#116946]
If a heap type does not have Py_TPFLAGS_HAVE_GC or doesn't not implement tp_traverse, the garbage collector may fail to collect the type *and* most of its instances at Python exit. The problem is worse when sub-interpreters are involved: if a sub-interpreter exits and does not clear all of its memory, the other interpreters will continue running with these leaks.

See issue gh-87138 for the rationale.

I wrote an article on the lock type which had this bug and caused a memory leak: https://vstinner.github.io/subinterpreter-leaks.html

### Linked PRs
* gh-116947
* gh-138290
* gh-138327
* gh-138328
* gh-138331
* gh-138333
* gh-138427
* gh-138428
* gh-138334
* gh-138340
* gh-138341
* gh-138582
* gh-138807
* gh-139073
* gh-139474
* gh-141268
* gh-141448

The following PRs were directly closed because the types are immutable and empty:

* gh-138329
* gh-138330
* gh-138332

The following PRs were wrongly merged so we reverted them:

* gh-138266
* gh-138320
* gh-138322
* gh-138288
* gh-138321
* gh-138323
* gh-138289
* gh-138324
* gh-138326

PRs for reverting the above commits:
 
* gh-138338
* gh-138337

**Comment by vstinner** [comment]
Attached PR gh-116947 should help to discover heap types which don't implement traverse.

For example, the `make` command lists 52 heap types which don't implement tp_traverse:

* _blake2.blake2b
* _blake2.blake2s
* _bz2.BZ2Compressor
* _bz2.BZ2Decompressor
* _curses_panel.panel
* functools._lru_list_elem
* _hashlib.HASH
* _hashlib.HASHXOF
* _hashlib.HMAC
* _lzma.LZMACompressor
* _lzma.LZMADecompressor
* posix.DirEntry
* posix.ScandirIterator
* _random.Random
* select.epoll
* select.poll
* _sha3.sha3_224
* _sha3.sha3_256
* _sha3.sha3_384
* _sha3.sha3_512
* _sha3.shake_128
* _sha3.shake_256
* _ssl.Certificate
* _testcapi.HeapCType
* _testcapi.HeapCTypeSetattr
* _testcapi.HeapCTypeSubclass
* _testcapi.HeapCTypeSubclassWithFinalizer
* _testcapi.HeapCTypeWithBuffer
* _testcapi.HeapCTypeWithDict
* _testcapi.HeapCTypeWithDict2
* _testcapi.HeapCTypeWithNegativeDict
* _testcapi.HeapCTypeWithWeakref
* _testcapi.HeapCTypeWithWeakref2
* _testcapi.HeapDocCType
* _testcapi.NullTpDocType
* _testcapi._test_structmembersType_NewAPI
* _testimportexec.Str
* _testlimitedcapi.LimitedVectorCallClass
* _thread._localdummy
* _thread._ThreadHandle
* _tkinter.Tcl_Obj
* _tkinter.tkapp
* _tkinter.tktimertoken
* _tokenize.TokenizerIter
* _xxinterpchannels.ChannelID
* xxlimited_35.Null
* xxlimited_35.Str
* xxlimited.Str
* _xxsubinterpreters.CrossInterpreterBufferView
* zlib.Compress
* zlib.Decompress
* zlib._ZlibDecompressor

**Comment by picnixz** [comment]
I use the following convention:

- When a list item has been addressed, I add a checkbox.
- When the behavior has been changed in main, I add :white_check_mark:.
- If nothing needs to be done, I cross the name.

The following items may need to support the GC protocol or be made immutable. Once this is done, this first list must be empty and each of its member will be moved to one of the lists below (may be present in more than one list).
 
- [ ] _interpreters.CrossInterpreterBufferView (?)

The following types are true container types that should support the GC protocol and have reference cycles with public documented API only. In particular, the corresponding PRs need to be backported.

- [x] :white_check_mark: _curses_panel.panel
- [x] :white_check_mark: _tkinter.tkapp [^1]
- [x] :white_check_mark: _tkinter.tktimertoken [^1]
- [x] :white_check_mark: zlib.Compress
- [x] :white_check_mark: zlib.Decompress
- [x] :white_check_mark: zlib._ZlibDecompressor

[^1]: Backports delayed until https://github.com/python/cpython/issues/138791 is fixed.

The following types are true container types that should support the GC protocol. This list tracks, in addition, their immutability in main (if this is not already the case):

- [x] :white_check_mark: _curses_panel.panel
- [x] :white_check_mark: _tkinter.tkapp
- [x] :white_check_mark: _tkinter.tktimertoken
- [x] :white_check_mark: zlib.Compress
- [x] :white_check_mark: zlib.Decompress

The following types are empty (or their attributes cannot be used to create cycles) and can be made immutable in main. Non-immutable trivial types are left as is in bugfix versions to alleviate the burden on the GC even if it's possible to create reference cycles manually (and require the user to dig into implementation details):

- [x] :white_check_mark: posix.DirEntry
- [x] :white_check_mark: posix.ScandirIterator
- [x] :white_check_mark: _random.Random 
- [x] :white_check_mark: _remote_debugging.RemoteUnwinder
- [x] :white_check_mark: select.epoll
- [x] :white_check_mark: select.poll
- [x] :white_check_mark: _tkinter.Tcl_Obj

The following types are immutable and empty (or cannot create cycles), hence do not need the GC. If they do implement the GC protocol, they should be simplified.

- [ ] _bz2.BZ2Compressor
- [ ] _bz2.BZ2Decompressor
- [ ] _hashlib.HASH
- [ ] _hashlib.HASHXOF
- [ ] _hashlib.HMAC
- [ ] _interpchannels.ChannelID
- [ ] _lzma.LZMACompressor
- [ ] _lzma.LZMADecompressor
- [ ] _ssl.Certificate
- [ ] _thread._localdummy
- [ ] _tokenize.TokenizerIter

The following types are immutable but not empty, but really internal.
Visiting them is delegated to someone else:

- [x] ✅ ~~functools._lru_list_elem~~ 

Finally, the following types are explicitly required to be missing the GC (testing purposes):

- [x] ✅ ~~_testcapi.HeapCTypeSetattr~~
- [x] ✅ ~~_testcapi.HeapCType~~
- [x] ✅ ~~_testcapi.HeapCTypeSubclass~~
- [x] ✅ ~~_testcapi.HeapCTypeSubclassWithFinalizer~~ (inherits `_testcapi.HeapCTypeSubclass`)

The following *should* require the GC protocol. Currently the tests do not leak, but they could if we were to store the instance as one of its attribute. Nevertheless, since we're only testing them once, I suggest we leave them alone.

- [ ] _testcapi.HeapCTypeWithDict2
- [ ] _testcapi.HeapCTypeWithDict
- [ ] _testcapi.HeapCTypeWithNegativeDict
- [ ] _testcapi.HeapCTypeWithRelativeDict

The following testing types are also empty or trivial:

- [x] ✅ ~~_testcapi.HeapCTypeVectorcall~~
- [x] ✅ ~~_testcapi.HeapCTypeWithBuffer~~
- [x] ✅ ~~_testcapi.HeapDocCType~~
- [x] ✅ ~~_testcapi.NullTpDocType~~
- [x] ✅ ~~_testcapi.PyUnicodeWriter~~
- [x] ✅ ~~_testcapi._test_structmembersType_NewAPI~~
- [x] ✅ ~~_testimportexec.Str~~
- [x] ✅ ~~_testlimitedcapi.LimitedRelativeVectorCallClass~~
- [x] ✅ ~~_testlimitedcapi.LimitedVectorCallClass~~
- [x] ✅ ~~xxlimited_35.Null~~
- [x] ✅ ~~xxlimited_35.Str~~
- [x] ✅ ~~xxlimited.Str~~

The following testing types are non-empty but do not own the references:

- [x] ✅ ~~_testcapi.HeapCTypeWithRelativeWeakref~~
- [x] ✅ ~~_testcapi.HeapCTypeWithWeakref2~~
- [x] ✅ ~~_testcapi.HeapCTypeWithWeakref~~

**Comment by picnixz** [comment]
So, according to https://github.com/python/cpython/pull/138329#issuecomment-3242122981, we should only implement the GC protocol if:

- the type is a true container (not just a thin wrapper).
- the type is a thin wrapper that is **not** an immutable type.

@kumaraditya303 am I right here? I assume that an empty type that **cannot** be instantiated, nor subclassed, but is mutable still needs to implement the GC protocol right?

**Comment by kumaraditya303** [comment]
> am I right here? I assume that an empty type that cannot be instantiated, nor subclassed, but is mutable still needs to implement the GC protocol right?

That's right, if it is mutable then it needs to be GC type regardless of being empty.

**Comment by serhiy-storchaka** [comment]
Many of this became possible after converting non-heap types to heap types. Non-heap types are immutable, so if the instance does not own references to other objects, and does not have mutable `__dict__`, it cannot create a reference loop.

I think that it would be better to make such types immutable again. This was not intentional change, but an side effect of conversion to heap types.

**Comment by picnixz** [comment]
Should we make those types immutable even for bugfix versions or should we also backport the immutability to bugfix versions? in the first case, I could retarget branches so that we only target 3.13 and 3.14 and create a different PR for main where we would make them immutable. In the second case, I would just modify the current PR.

Note: most non-immutable heap types were introduced in Python 3.9.

**Comment by serhiy-storchaka** [comment]
It may be dangerous to add Py_TPFLAGS_HAVE_GC in some version and then remove it in the next version. I think it is safer to leave the original bug (possibility of creating a reference loop) unfixed for that types. It is very unlikely to occur in real world code -- you need to modify the type which is supposed to be immutable.

I only talk about types whose instances can't contain strong references to arbitrary objects.

**Comment by picnixz** [comment]
Ok, then we should add the traverse functions where it's needed then.

**Comment by serhiy-storchaka** [comment]
Unneeded Py_TPFLAGS_HAVE_GC just adds more work for garbage collector (and increases memory consumption). I would prefer to not add it without need. Just add Py_TPFLAGS_IMMUTABLETYPE. It did not exist in 3.9, but it was used for most conversion to heap types since 3.10.

**Comment by picnixz** [comment]
I still don't know what to do with  _interpreters.CrossInterpreterBufferView. It's a valid base type that is immutable, though.

**Comment by kumaraditya303** [comment]
There are a number of heap types which are currently unnecessary gc types like `_thread.lock`, it would be nice to remove that as it would reduce memory usage and reduce pressure on gc, in case anyone wants to work on it.

**Comment by sergey-miryanov** [comment]
I want to work on this if anyone don't mind.

**Comment by picnixz** [comment]
Yeah go ahead Sergey. Check which ones don't need the traverse functions.

**Comment by sergey-miryanov** [comment]
> There are a number of heap types which are currently unnecessary gc types like `_thread.lock`, it would be nice to remove that as it would reduce memory usage and reduce pressure on gc, in case anyone wants to work on it.

@kumaraditya303 There is an existing bug about having of `Py_TPFLAGS_MANAGED_WEAKREF` without `Py_TPFLAGS_HAVE_GC` one - https://github.com/python/cpython/issues/134786. I worked on it before, but then postponed it. Now, I'm planning to fix it. Afterwards, I will return to this issue.

`_thread.lock` has both flags. So if I remove `Py_TPFLAGS_HAVE_GC` then python crashes.

**Comment by picnixz** [comment]
@sergey-miryanov In the list above, there are some types that are empty and immutable but still have the GC (the list contains BZ2, LZMA and crypto types). I can go through each of them to check which ones can be simplified; otherwise, I'll let you work on this.

**Comment by sergey-miryanov** [comment]
@picnixz Ok, I will check the list first. Thanks.

**Comment by sergey-miryanov** [comment]
@picnixz Should I add a separate PR for each module or add only one with all changes? Prior PRs were as one PR per module, IIUC.

**Comment by picnixz** [comment]
If possible, I would prefer one big PR here as we won't need backports and thjs reverting will be easier. See the PR I made when making types immutable. I however don't think you'll need any NEWS entry as this won't be a visible change.

**Comment by sergey-miryanov** [comment]
- [x] _bz2.BZ2Compressor
- [x] _bz2.BZ2Decompressor
- [ ] _hashlib.HASH - **do not need**
- [ ] _hashlib.HASHXOF - **do not need**
- [x] _hashlib.HMAC
- [ ] _interpchannels.ChannelID - **do not need**
- [x] _lzma.LZMACompressor
- [x] _lzma.LZMADecompressor
- [ ] _ssl.Certificate - **do not need**
- [ ] _thread._localdummy - **do not need**
- [ ] _tokenize.TokenizerIter - **do not need**

Also fixed:
- [x] _ssl.MemoryBIO
- [x] _thread._ThreadHandle

**Comment by sergey-miryanov** [comment]
@picnixz I have added a PR, please take a look.

**Comment by neonene** [comment]
In this discussion, heaptypes created with `Py_TPFLAGS_IMMUTABLETYPE` are being considered to not require `Py_TPFLAGS_HAVE_GC`, right?  If so, I'm wondering why `_thread._ThreadHandle` (immutable heaptype) caused ref leaks without GC support in #116915.

**Comment by picnixz** [comment]
That's weird. Could it be because we are indirectly creating a cycle because of the inner thread handle? I now am really confused because sometimes we do have leaks and sometimes we do not.

Kumar is away for a month but @serhiy-storchaka should be around. I would like to know what exactly happens in this case. For the other empty immutable types I couldn't reproduce a leak so I don't know if in the thread issue, we correctly diagnosed the root of the problem.

**Comment by kumaraditya303** [comment]
> I'm wondering why _thread._ThreadHandle (immutable heaptype) caused ref leaks without GC support in https://github.com/python/cpython/issues/116915.

That wasn't a real refleak and it occurred because register_at_fork was called, which the refleak checker mistakenly identifies as a leak. This is a known, longstanding issue in CPython that fork handlers cannot be cleared.

See this comment https://github.com/python/cpython/blob/ce23eeaa2b3f3cea175779791726424ade69a8ef/Lib/test/test_posix.py#L243-L245

**Comment by vfazio** [comment]
I'm in the process of trying to convert a single phase init module (https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/tree/bindings/python/tests/gpiosim/ext.c) over to multi-phase init in light of PEP 793 looking to soft deprecate PyInit_*. 

I'm trying to keep up with the back and forth on `Py_TPFLAGS_IMMUTABLETYPE` and types created via `PyType_FromModuleAndSpec` which are `Py_TPFLAGS_HEAPTYPE`.

In our case, the type does need access to module state, so needs to be a heap type after the transition.

What's not clear is whether I also need to define this as `Py_TPFLAGS_HAVE_GC` and define `tp_traverse`. It looks like the standard modules like `bz2` (which is probably our closest analogue) has been marked GC, unmarked GC, and maybe still under consideration to be added back https://github.com/python/cpython/pull/139073#issuecomment-3308545866

If I'm reading https://github.com/python/cpython/pull/118021#discussion_r1572062702 correctly, if a heap type is created via `PyType_FromSpec` and thus has no module ref, it may be possible to avoid the GC flag, otherwise if using methods like `PyType_FromModuleAndSpec` or `PyType_FromMetaclass` with a module argument, then there may be a need to add the GC flag and a `tp_traverse` method?

Pep 573 https://peps.python.org/pep-0573/#adding-module-references-to-heap-types:
> Usually, creating a class with ht_module set will create a reference cycle involving the class and the module. This is not a problem, as tearing down modules is not a performance-sensitive operation, and module-level functions typically also create reference cycles. The existing “set all module globals to None” code that breaks function cycles through f_globals will also break the new cycles through ht_module.

Have there been any new discoveries or is there any clarification on what's recommended for C extension maintainers? Just looking for some guidance, otherwise I may just punt on converting this until this issue has been resolved.

**Comment by serhiy-storchaka** [comment]
PR #152310 reimplements the GC protocol for `tkapp` and `tktimertoken`, addressing the two problems that led to the gh-138331 revert (gh-138807):

- The shutdown crash (gh-138789): the GC slots use a plain cast instead of the type-checking macro, since the collector can visit a surviving object after `module_clear()` has reset the global type pointers.
- The cancelled-timer regression: a pending timer is kept alive by the Tcl event loop, so it is treated as a GC root (only its callback is traversed) and collecting it no longer cancels a live timer.

Both are covered by regression tests.

**Comment by skirpichev** [comment]
> Have there been any new discoveries or is there any clarification on what's recommended for C extension maintainers?

I think we should create a separate documentation issue.  @serhiy-storchaka ?

As currently we have [following advice](https://docs.python.org/3.16/howto/isolating-extensions.html#garbage-collection-protocol) in docs:
> Instances of heap types hold a reference to their type. This ensures that the type isn’t destroyed before all its instances are, but may result in reference cycles that need to be broken by the garbage collector.
>
> To avoid memory leaks, instances of heap types must implement the garbage collection protocol. That is, heap types should:
>
>    Have the [Py_TPFLAGS_HAVE_GC](https://docs.python.org/3.16/c-api/typeobj.html#c.Py_TPFLAGS_HAVE_GC) flag.
>
>    Define a traverse function using [Py_tp_traverse](https://docs.python.org/3.16/c-api/typeobj.html#c.Py_tp_traverse), which visits the type (e.g. using Py_VISIT(Py_TYPE(self))).

If that's actually optional, I believe it worth to be documented.  Using GC-enabled immutable heap types come with a noticeable performance loss.  Here is an example for the `_decimal` module: https://github.com/skirpichev/cpython/pull/24
